kv/kvserver: TestQuotaPool failed

[(kv/kvserver).TestQuotaPool failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2613496&tab=buildLog) on [master@fad50c1b65515bfa11c12f006e79dc4d44d02e0b](https://github.com/cockroachdb/cockroach/commits/fad50c1b65515bfa11c12f006e79dc4d44d02e0b):

```
=== RUN   TestQuotaPool
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQuotaPool722528802
    test_log_scope.go:74: use -show-logs to present logs inline
    client_raft_test.go:2299: didn't observe the expected quota acquisition, available: 10000
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQuotaPool722528802
--- FAIL: TestQuotaPool (0.77s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestQuotaPool PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQuotaPool.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestQuotaPool failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2614366&tab=buildLog) on [master@05a1ab60cf15a9326a93fddfff70d7ff59005e3e](https://github.com/cockroachdb/cockroach/commits/05a1ab60cf15a9326a93fddfff70d7ff59005e3e):

```
=== RUN   TestQuotaPool
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQuotaPool638784974
    test_log_scope.go:74: use -show-logs to present logs inline
    client_raft_test.go:2299: didn't observe the expected quota acquisition, available: 10000
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQuotaPool638784974
--- FAIL: TestQuotaPool (1.01s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestQuotaPool PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQuotaPool.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -2187,7 +2187,6 @@ func TestReplicateAddAndRemove(t *testing.T) {
 // to constantly catch up the slower node via snapshots. See #8659.
 func TestQuotaPool(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 59382, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	const quota = 10000
@@ -2208,29 +2207,8 @@ func TestQuotaPool(t *testing.T) {
 		})
 	defer tc.Stopper().Stop(ctx)
 
-	key := []byte("a")
-	tc.SplitRangeOrFatal(t, key)
+	key := tc.ScratchRange(t)
 	tc.AddVotersOrFatal(t, key, tc.Targets(1, 2)...)
-
-	assertEqualLastIndex := func() error {
-		var expectedIndex uint64
-		for i := range tc.Servers {
-			repl := tc.GetFirstStoreFromServer(t, i).LookupReplica(key)
-			require.NotNil(t, repl)
-
-			index, err := repl.GetLastIndex()
-			if err != nil {
-				t.Fatal(err)
-			}
-			if i == 0 {
-				expectedIndex = index
-			} else if expectedIndex != index {
-				return fmt.Errorf("%s: expected lastIndex %d, but found %d", repl, expectedIndex, index)
-			}
-		}
-		return nil
-	}
-	testutils.SucceedsSoon(t, assertEqualLastIndex)
 
 	// NB: See TestRaftBlockedReplica/#9914 for why we use a separate	goroutine.
 	raftLockReplica := func(repl *kvserver.Replica) {
@@ -2239,7 +2217,7 @@ func TestQuotaPool(t *testing.T) {
 		<-ch
 	}
 
-	leaderRepl := tc.GetRaftLeader(t, key)
+	leaderRepl := tc.GetRaftLeader(t, roachpb.RKey(key))
 	// Grab the raftMu to re-initialize the QuotaPool to ensure that we don't
 	// race with ongoing applications.
 	raftLockReplica(leaderRepl)
@@ -2247,9 +2225,24 @@ func TestQuotaPool(t *testing.T) {
 		t.Fatalf("failed to initialize quota pool: %v", err)
 	}
 	leaderRepl.RaftUnlock()
+	// Wait until the follower will is not ignored by updateProposalQuotaRaftMuLocked.
+	// Otherwise the quota gets prematurely released into the pool. We can use status.Applied
+	// instead of proposalQuotaBaseIndex because it's strictly ahead or the same.
+	testutils.SucceedsSoon(t, func() error {
+		var err error
+		status := leaderRepl.RaftStatus()
+		minIndex := status.Applied
+		for id, progress := range status.Progress {
+			if progress.Match < minIndex {
+				err = errors.Errorf("Replica %d is behind leader expected %d but was %d", id, minIndex, progress.Match)
+			}
+		}
+		return err
+	})
+
 	followerRepl := func() *kvserver.Replica {
 		for i := range tc.Servers {
-			repl := tc.GetFirstStoreFromServer(t, i).LookupReplica(key)
+			repl := tc.GetFirstStoreFromServer(t, i).LookupReplica(roachpb.RKey(key))
 			require.NotNil(t, repl)
 			if repl == leaderRepl {
 				continue
@@ -2281,10 +2274,10 @@ func TestQuotaPool(t *testing.T) {
 		// second write, previously blocked by virtue of there not being enough
 		// quota, is now free to proceed. We expect the final quota in the system
 		// to be the same as what we started with.
-		key := roachpb.Key("k")
+		keyToWrite := key.Next()
 		value := bytes.Repeat([]byte("v"), (3*quota)/4)
 		var ba roachpb.BatchRequest
-		ba.Add(putArgs(key, value))
+		ba.Add(putArgs(keyToWrite, value))
 		if err := ba.SetActiveTimestamp(tc.Servers[0].Clock().Now); err != nil {
 			t.Fatal(err)
 		}
@@ -2305,7 +2298,7 @@ func TestQuotaPool(t *testing.T) {
 
 		go func() {
 			var ba roachpb.BatchRequest
-			ba.Add(putArgs(key, value))
+			ba.Add(putArgs(keyToWrite, value))
 			if err := ba.SetActiveTimestamp(tc.Servers[0].Clock().Now); err != nil {
 				ch <- roachpb.NewError(err)
 				return

